File: macos/Sources/SafeFileReader.swift
import Foundation

enum SafeFileReader {
    private static let maxAttempts = 5

    static func readData(from url: URL) throws -> Data {
        guard FileManager.default.fileExists(atPath: url.path) else {
            throw NSError(domain: NSCocoaErrorDomain, code: NSFileReadNoSuchFileError, userInfo: [
                NSLocalizedDescriptionKey: "File not found."
            ])
        }

        var lastError: Error?
        for attempt in 0..<maxAttempts {
            if attempt > 0 {
                Thread.sleep(forTimeInterval: 0.05 * Double(attempt))
            }

            do {
                let sizeBefore = try FileManager.default.attributesOfItem(atPath: url.path)[.size] as? Int64 ?? 0
                let data = try Data(contentsOf: url)
                let sizeAfter = try FileManager.default.attributesOfItem(atPath: url.path)[.size] as? Int64 ?? 0
                if Int64(data.count) == sizeBefore, Int64(data.count) == sizeAfter {
                    return data
                }
            } catch {
                lastError = error
            }
        }

        if let lastError {
            throw NSError(domain: "TextPad", code: 6, userInfo: [
                NSLocalizedDescriptionKey: "Unable to read a stable copy of \"\(url.lastPathComponent)\".",
                NSUnderlyingErrorKey: lastError
            ])
        }

        return try Data(contentsOf: url)
    }
}

File: windows/TextPad/Services/SafeFileReader.cs
using System.IO;

namespace TextPad.Services;

public static class SafeFileReader
{
    private const int MaxAttempts = 5;

    public static byte[] ReadAllBytes(string path)
    {
        if (!File.Exists(path))
            throw new FileNotFoundException("File not found.", path);

        Exception? lastError = null;
        for (var attempt = 0; attempt < MaxAttempts; attempt++)
        {
            if (attempt > 0)
                Thread.Sleep(50 * attempt);

            try
            {
                var sizeBefore = new FileInfo(path).Length;
                var bytes = File.ReadAllBytes(path);
                var sizeAfter = new FileInfo(path).Length;
                if (bytes.Length == sizeBefore && bytes.Length == sizeAfter)
                    return bytes;
            }
            catch (IOException ex)
            {
                lastError = ex;
            }
            catch (UnauthorizedAccessException ex)
            {
                lastError = ex;
            }
        }

        throw new IOException(
            $"Unable to read a stable copy of \"{path}\".",
            lastError ?? new IOException("The file changed while it was being read."));
    }
}